hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/ContainerClientMetrics.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hdds.scm;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.hdds.scm.pipeline.Pipeline;
+import org.apache.hadoop.hdds.scm.pipeline.PipelineID;
+import org.apache.hadoop.metrics2.annotation.Metric;
+import org.apache.hadoop.metrics2.annotation.Metrics;
+import org.apache.hadoop.metrics2.lib.DefaultMetricsSystem;
+import org.apache.hadoop.metrics2.lib.Interns;
+import org.apache.hadoop.metrics2.lib.MetricsRegistry;
+import org.apache.hadoop.metrics2.lib.MutableCounterLong;
+import org.apache.hadoop.ozone.OzoneConsts;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Container client metrics that describe how data writes are distributed to
+ * pipelines.
+ */
+@Metrics(about = "Client Metrics", context = OzoneConsts.OZONE)
+public final class ContainerClientMetrics {
+  private static ContainerClientMetrics instance;
+  private static int referenceCount = 0;
+
+  private static final String SOURCE_NAME =
+      ContainerClientMetrics.class.getSimpleName();
+
+  @Metric
+  private MutableCounterLong totalWriteChunkCalls;
+  @Metric
+  private MutableCounterLong totalWriteChunkBytes;
+  private final Map<PipelineID, MutableCounterLong> writeChunkCallsByPipeline;
+  private final Map<PipelineID, MutableCounterLong> writeChunkBytesByPipeline;
+  private final Map<UUID, MutableCounterLong> writeChunksCallsByLeaders;
+  private final MetricsRegistry registry;
+
+  public static synchronized ContainerClientMetrics acquire() {
+    if (instance == null) {
+      instance = newInstance();
+    }
+    referenceCount++;
+    return instance;
+  }
+
+  @VisibleForTesting
+  public static ContainerClientMetrics newInstance() {
+    return DefaultMetricsSystem.instance().register(SOURCE_NAME,
+        "Ozone Client Metrics", new ContainerClientMetrics());
```

Comment:
It seems metrics system expects name to be unique, even after unregister/register cycle.  We can achieve that by adding an instance counter and generating name as `SOURCE_NAME + counter` when creating new instance, and then using it in `release()`, too.

Reply:
I think the name can be reused after being unregistered. Added a unit test to verify.

Reply:
Problem is specific to Hadoop 2.7, our Hadoop 3.3 MR test pass with the same code.  I think this was a Hadoop bug, fixed in [YARN-5190](https://issues.apache.org/jira/browse/YARN-5190).

> 1. Fix the asymmetric behaviors in register()/unregisterSource() at MetricsSystemImpl that source name is still left in sourceNames.map in DefaultMetricsSystem after unregisterSource().

https://github.com/apache/hadoop/commit/99cc439e29794f8e61bebe03b2a7ca4b6743ec92#diff-299fd2f06692c12bf10a572a87be5010de9c4cd2563a3d366ba0d9b76f038d3a

Reply:
Got it, thanks.